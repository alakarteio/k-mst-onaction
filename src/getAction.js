import { getPath } from 'mobx-state-tree'

export default (call) => {
  const {
    name,
    type,
    context,
    args,
  } = call

  const path = getPath(context)
  const fullpath = `${path}/${name}`

  const action = {
    fullpath,
    path,
    name,
    args,
  }

  switch (type) {
    case 'process_return': // mobx-state-tree v1.0 and older
    case 'flow_return': return { ...action, ended: true }
    case 'action': return action
    default: return undefined
  }
}
