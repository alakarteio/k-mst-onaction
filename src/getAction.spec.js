/* eslint-env jest */
import getAction from './getAction'

describe('getAction maper', () => {
  it('should map a simple mobx-state-tree action', () => {
    const call = {
      name: 'actionName',
      type: 'action', // simple mobx-state-tree action
      context: { $treenode: { path: '/this/is/a/path' } },
      args: ['first', { second: 'argument' }],
    }

    const res = getAction(call)

    expect(res).toMatchSnapshot()
  })

  it('should map an undefined type', () => {
    const call = {
      name: 'actionName',
      type: 'undefined type',
      context: { $treenode: { path: '/this/is/a/path' } },
      args: ['first', { second: 'argument' }],
    }

    const res = getAction(call)

    expect(res).toMatchSnapshot()
  })

  it('should map an asynchronous action that ends', () => {
    const call = {
      name: 'actionName',
      type: 'process_return', // asynchronous that ends action
      context: { $treenode: { path: '/this/is/a/path' } },
      args: ['first', { second: 'argument' }],
    }

    const res = getAction(call)

    expect(res).toMatchSnapshot()
  })
})
