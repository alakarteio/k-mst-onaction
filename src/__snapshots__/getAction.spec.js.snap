// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAction maper should map a simple mobx-state-tree action 1`] = `
Object {
  "args": Array [
    "first",
    Object {
      "second": "argument",
    },
  ],
  "fullpath": "/this/is/a/path/actionName",
  "name": "actionName",
  "path": "/this/is/a/path",
}
`;

exports[`getAction maper should map an asynchronous action that ends [mst<=1.0.0] 1`] = `
Object {
  "args": Array [
    "first",
    Object {
      "second": "argument",
    },
  ],
  "ended": true,
  "fullpath": "/this/is/a/path/actionName",
  "name": "actionName",
  "path": "/this/is/a/path",
}
`;

exports[`getAction maper should map an asynchronous action that ends [mst>1.0.0] 1`] = `
Object {
  "args": Array [
    "first",
    Object {
      "second": "argument",
    },
  ],
  "ended": true,
  "fullpath": "/this/is/a/path/actionName",
  "name": "actionName",
  "path": "/this/is/a/path",
}
`;

exports[`getAction maper should map an undefined type 1`] = `undefined`;
